File: Try_GameEngine5/src/com/example/try_gameengine/action/InverseMoveOrderDecorator.java
package com.example.try_gameengine.action;

import java.util.Collections;
import java.util.List;

import android.util.Log;

public class InverseMoveOrderDecorator extends MovementDecorator {
	private MovementAction action;

	public InverseMoveOrderDecorator(MovementAction action) {
		this.action = action;
		this.copyMovementActionList = action.copyMovementActionList;
	}

	private MovementActionInfo coreCalculationMovementActionInfo(
			MovementActionInfo info) {
		return info;
	}

	@Override
	public void start() {
		action.getAction().start();
	}

	@Override
	public MovementAction getAction() {
		return action.getAction();
	}

	@Override
	public String getDescription() {
		return "Double " + action.getDescription();
	}

	@Override
	public MovementAction initMovementAction() {
		return initTimer();
	}

	@Override
	protected MovementAction initTimer() {

		if (this.getAction().getActions().size() == 0) {
			action.getAction().setInfo(getInfo());
			action.getAction().initTimer();
		} else {
			this.getAction().initTimer();
			doIn();
		}
		return this;
	}

	@Override
	public MovementAction addMovementAction(MovementAction action) {
		getAction().addMovementAction(action);
		return this;
	}

	@Override
	protected void setActionsTheSameTimerOnTickListener() {
		getAction().setTimerOnTickListener(timerOnTickListener);
	}

	@Override
	public MovementActionInfo getInfo() {
		return coreCalculationMovementActionInfo(action.getInfo());
	}

	@Override
	public List<MovementAction> getCurrentActionList() {
		// TODO Auto-generated method stub
		return action.getCurrentActionList();
	}

	@Override
	public List<MovementActionInfo> getCurrentInfoList() {
		// TODO Auto-generated method stub
		return action.getCurrentInfoList();
	}

	@Override
	public List<MovementAction> getMovementItemList() {
		return action.getMovementItemList();
	}

	@Override
	public List<MovementActionInfo> getMovementInfoList() {
		return action.getMovementInfoList();
	}

	@Override
	public void doIn() {
		action.doIn();
		this.getAction().getCurrentInfoList();
		int i = 0;
		for (MovementActionInfo info : this.getAction().currentInfoList) {
			Log.e("count", ++i + "");
			Log.e("info", info.getDx() + "");
			this.getAction().setInfo(info);
			coreCalculationMovementActionInfo(this.getAction().getInfo());
		}

		inverseOrder(this);

		for (MovementAction movementItem : this.getAction().movementItemList) {
			movementItem.initTimer();
		}
	}

	private void inverseOrder(MovementAction targetAction) {
		Collections.reverse(targetAction.getAction().getActions());
		for (MovementAction action : targetAction.getAction().getActions()) {
			inverseOrder(action);
		}
	}
}

File: Try_GameEngine5/src/com/example/try_gameengine/action/InverseMoveOrderDecorator.kt
package com.example.try_gameengine.action

import java.util.Collections

class InverseMoveOrderDecorator(action: MovementAction) : MovementDecorator() {
    init {
        this.action = action
        //		this.copyMovementActionList = action.copyMovementActionList;
    }

    public override fun coreCalculationMovementActionInfo(
        action: MovementAction
    ): MovementAction? {
        return action
    }

    override fun start() {
        action.getAction().start()
    }

    override fun getAction(): MovementAction {
        return action.getAction()
    }

    override fun getDescription(): String {
        return "Double " + action.getDescription()
    }

    public override fun initTimer(): MovementAction {
        super.initTimer()

        if (this.getAction()!!.getActions().size == 0) {
            action.getAction().setInfo(getInfo())
            action.getAction().initTimer()
        } else {
            this.getAction()!!.initTimer()
            //			doIn(null);
        }
        return this
    }

    override fun addMovementAction(action: MovementAction): MovementAction {
        getAction()!!.addMovementAction(action)
        return this
    }

    override fun setActionsTheSameTimerOnTickListener() {
        getAction()!!.setTimerOnTickListener(timerOnTickListener)
    }

    override fun getCurrentActionList(): MutableList<MovementAction> {
        // TODO Auto-generated method stub
        return action.getCurrentActionList()
    }

    override fun getCurrentInfoList(): MutableList<MovementActionInfo?> {
        // TODO Auto-generated method stub
        return action.getCurrentInfoList()
    }

    override fun getMovementInfoList(): MutableList<MovementActionInfo?> {
        return action.getMovementInfoList()
    }

    public override fun doIn(actionSet: MovementActionSet?): MutableList<MovementAction> {
        val actions = super.doIn(actionSet)

        //		this.getAction().getCurrentInfoList();
//		int i = 0;
//		for (MovementActionInfo info : this.getAction().currentInfoList) {
//			Log.e("count", ++i + "");
//			Log.e("info", info.getDx() + "");
//			this.getAction().setInfo(info);
        // /**/            coreCalculationMovementActionInfo(this.getAction()!!.getInfo()); */
//			coreCalculationMovementActionInfo(this.getAction());
//		}
        inverseOrder(this)

        return actions

        //		for (MovementAction movementItem : this.getAction().movementItemList) {
//			movementItem.initTimer();
//		}
    }

    private fun inverseOrder(targetAction: MovementAction) {
        Collections.reverse(targetAction.getAction().getActions())
        for (action in targetAction.getAction().getActions()) {
            inverseOrder(action)
        }
    }
}
